firedrake/assign.py

```diff
@@ -0,0 +1,130 @@
+import abc
+
+import numpy as np
+from pyop2.utils import cached_property
+import ufl
+from ufl.algorithms import extract_coefficients
+from ufl.constantvalue import RealValue, Zero
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction
+
+from firedrake.utils import ScalarType
+from firedrake.vector import Vector
+
+
+class WeightedCoefficientCollector(MultiFunction):
+    def __init__(self):
+        super().__init__()
+        self.weighted_coefficients = []
+
+    def sum(self, o):
+        for op in o.ufl_operands:
+            self(op)
+
+    def product(self, o):
+        try:
+            weight, = filter(
+                lambda op: isinstance(op, (RealValue, Zero)), o.ufl_operands
+            )
+            coeff, = filter(lambda op: isinstance(op, ufl.Coefficient), o.ufl_operands)
+        except ValueError:
+            raise ValueError("Only products of scalars and coefficients allowed")
+        self.weighted_coefficients.append((weight, coeff))
+
+    def coefficient(self, o):
+        self.weighted_coefficients.append((1, o))
```

Comment:
Prefer to return values and glue them together coming back up the traversal rather than side-effectful modification of the visitor. That way you will be able to get away with only constructing a single `WeightedCoefficientCollector` (rather than one for every expression).